components/brave_wallet/browser/nft_metadata_fetcher.cc

```diff
@@ -0,0 +1,376 @@
+  std::move(callback).Run(response, mojom::ProviderError(error), error_message);
+}
+
+void NftMetadataFetcher::GetSolTokenMetadata(
+    const std::string& nft_account_address,
+    GetSolTokenMetadataCallback callback) {
+  // Derive metadata PDA for the NFT accounts
+  absl::optional<std::string> associated_metadata_account =
+      SolanaKeyring::GetAssociatedMetadataAccount(nft_account_address);
+  if (!associated_metadata_account) {
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&NftMetadataFetcher::OnGetSolanaAccountInfoTokenMetadata,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+  json_rpc_service_->GetSolanaAccountInfo(*associated_metadata_account,
+                                          std::move(internal_callback));
+}
+
+void NftMetadataFetcher::OnGetSolanaAccountInfoTokenMetadata(
+    GetSolTokenMetadataCallback callback,
+    absl::optional<SolanaAccountInfo> account_info,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || !account_info) {
+    std::move(callback).Run("", std::move(error), error_message);
+    return;
+  }
+
+  absl::optional<std::vector<uint8_t>> metadata =
+      base::Base64Decode((*account_info).data);
+  if (!metadata) {
+    std::move(callback).Run("", mojom::SolanaProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  absl::optional<GURL> url = DecodeMetadataUri(*metadata);
+  if (!url || !url.value().is_valid()) {
+    std::move(callback).Run("", mojom::SolanaProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  FetchMetadata(*url, base::BindOnce(
+                          &NftMetadataFetcher::CompleteGetSolTokenMetadata,
+                          weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void NftMetadataFetcher::CompleteGetSolTokenMetadata(
+    GetSolTokenMetadataCallback callback,
+    const std::string& response,
+    int error,
+    const std::string& error_message) {
+  std::move(callback).Run(response, mojom::SolanaProviderError(error),
+                          error_message);
+}
+
+// static
+absl::optional<uint32_t> NftMetadataFetcher::DecodeUint32(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < sizeof(uint32_t)) {
+    return absl::nullopt;
+  }
+
+  // Read bytes in little endian order.
+  base::span<const uint8_t> s =
+      base::make_span(input.begin() + offset, sizeof(uint32_t));
+  uint32_t uint32_le = *reinterpret_cast<const uint32_t*>(s.data());
+
+  offset += sizeof(uint32_t);
+
+#if defined(ARCH_CPU_LITTLE_ENDIAN)
```

Comment:
Should we assume that the input value here is going to be little endian?

Reply:
I think it's reasonable to only support Borsh encoding from a product perspective.  I am not sure how commonly we will encounter other formats, if ever.

Reply:
Ok, I think that's reasonable as well. We can take a look at this in a separate refactor.